libs/elasticsearch-nio/src/main/java/org/elasticsearch/nio/ReadWriteHandler.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.nio;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.BiConsumer;
+
+/**
+ * Implements the application specific logic for handling inbound and outbound messages for a channel.
+ */
+public interface ReadWriteHandler {
+
+    /**
+     * This method is called when a message is queued with a channel. It can be called from any thread.
+     * This method should validate that the message is a valid type and return a write operation object
+     * to be queued with the channel
+     *
+     * @param context the channel context
+     * @param message the message
+     * @param listener the listener to be called when the message is sent
+     * @return the write operation to be queued
+     */
+    WriteOperation createWriteOperation(SocketChannelContext context, Object message, BiConsumer<Void, Throwable> listener);
```

Comment:
any chance we can make this typesafe?

Reply:
This is something I have thought about and I don't think it is very easy to do in this PR. The `ReadWriteHandler` would need to be parameterized. And then the channel context would need to be parameterized. And then the channel. 

I think down the line it might be more possible. With implementing an "http" channel type that is a subclass of a niochannel (similar to how we have a tcp channel type). But I think that would be a lot more code and better to do as follow-up work.

Reply:
++